### PR TITLE
internal/ethapi: optimize & clean up EstimateGas

### PR DESCRIPTION
**Pull request by roberto-bayardo**
Optimizations:
- Currently if a transaction is reverting, EstimateGas will exhibit worst-case behavior and binary search up to the max gas limit (~40 state-clone + tx executions).  This PR allows EstimateGas to return after only a single unconstrained execution in this scenario.
- Uses the gas used from the unconstrained execution to bias the remaining binary search towards the likely solution in a simple way that doesn't impact the worst case. For a typical contract-invoking transaction, this reduces the median number of state-clone+executions from 25 to 18 (28% reduction).

Cleanup:
- added & improved function + code comments
- correct the EstimateGas documentation to clarify the gas limit determination is at latest block, not pending, if the blockNr is unspecified.

The performance analysis was conducted over 2300 contract-invoking transactions randomly sampled (via tx hash starting with "00")  from the mempool over the past 24 hours. These txs were put through the new implementation & old implementation with the exact same state when the tx hit the mempool, and the # of executions for each recorded.   The output from both calls was also compared to ensure they were equivalent.  There were a few (<< 1%) instances where the values differed, upon further investigation these were due to transactions that monitor gas-remaining and attempt to adjust computation accordingly (such as batch NFT minting, [example](https://etherscan.io/address/0x0dE8bf93dA2f7eecb3d9169422413A9bef4ef628#code)).

### CODE REVIEW

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -1170,13 +1154,15 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
 	} else {
 		feeCap = common.Big0
 	}
+
+	st, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
```
Comment (roberto-bayardo) [inline internal/ethapi/api.go:1158]: had to use "st" instead of "state" for this var to avoid name conflict with type state.StateDB referenced below

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -1202,35 +1188,56 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
...
+	// Binary search for the smallest gas limit that allows the tx to execute successfully.
 	for lo+1 < hi {
-		s := state.Copy()
 		mid := (hi + lo) / 2
-		failed, _, err := executable(mid, s, header)
-
-		// If the error is not nil(consensus error), it means the provided message
-		// call or transaction will never be accepted no matter how much gas it is
-		// assigned. Return the error directly, don't struggle any more.
+		if mid > lo*2 {
```
Comment (roberto-bayardo) [inline internal/ethapi/api.go:1232]: There is likely room for further average-case improvement by tweaking (lowering) the constant multiplier here, but this would worsen the worst-case and also introduces new boundary conditions that would need to be checked.

File: internal/ethapi/api.go (hunk 1 above)
Comment (lightclient) [reply internal/ethapi/api.go:1158]: I'd probably just pull the `execute` function out into something like `executeEstimate` since you're giving up on expressiveness of variable name to appease the compiler.

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -1134,12 +1114,16 @@ func (s *BlockChainAPI) Call(ctx context.Context, args TransactionArgs, blockNrO
...
+// there are unexpected failures. The gas limit is capped by both `args.Gas` (if non-nil &
+// non-zero) and `gasCap` (if non-zero).
 func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash rpc.BlockNumberOrHash, gasCap uint64) (hexutil.Uint64, error) {
-	// Binary search the gas requirement, as it may be higher than the amount used
+	// Binary search the gas limit, as it may need to be higher than the amount used
 	var (
-		lo  uint64 = params.TxGas - 1
-		hi  uint64
-		cap uint64
+		// `lo` and `hi` will define the current range of our binary search
```
Comment (lightclient) [inline internal/ethapi/api.go:1124]:
```suggestion
```
self explanatory

File: internal/ethapi/api.go [hunk 4]
```diff
@@ -1202,35 +1188,56 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
...
+		}
+		return 0, fmt.Errorf("gas required exceeds allowance (%d)", hi)
+	}
+	// For almost any transaction, the gas consumed by the unconstrained execution above
+	// lower-bounds the gas limit required for it to succeed. One exception is those txs that
+	// explicitly check gas remaining in order to successfully execute within a given limit, but we
+	// probably don't want to return a lowest possible gas limit for these cases anyway.
+	lo = gasUsed - 1
+
+	// Binary search for the smallest gas limit that allows the tx to execute successfully.
```
Comment (lightclient) [inline internal/ethapi/api.go:1229]: Not really so much binary search if you adjust the mid point yeah?

File: internal/ethapi/api.go (hunk 2 above)
Comment (lightclient) [reply internal/ethapi/api.go:1232]: This is more of an acceptance that there are a much larger number of transactions on the low size, 21k - 500k gas?

File: internal/ethapi/api.go [hunk 5]
```diff
@@ -1062,17 +1063,17 @@ func doCall(ctx context.Context, b Backend, args TransactionArgs, state *state.S
...
 	if evm.Cancelled() {
-		return nil, fmt.Errorf("execution aborted (timeout = %v)", timeout)
+		return nil, 0, fmt.Errorf("execution aborted (timeout = %v)", timeout)
 	}
 	if err != nil {
-		return result, fmt.Errorf("err: %w (supplied gas %d)", err, msg.GasLimit)
+		return result, 0, fmt.Errorf("err: %w (supplied gas %d)", err, msg.GasLimit)
 	}
-	return result, nil
+	return result, math.MaxUint64 - gp.Gas(), nil
```
Comment (lightclient) [inline internal/ethapi/api.go:1076]: I don't think it is great using `math.MaxUint64` here with the requirement it is the starting point for the gas pool. But more importantly, I don't see why the function signature needs to change here. `result` has `gasUsed` as a field. Is this not an acceptable way of determining the amount of gas used on the call?

File: internal/ethapi/api.go (hunk 4 above)
Comment (roberto-bayardo) [reply internal/ethapi/api.go:1229]: It's a pure binary search after 1-2 doublings of 'lo' in the typical case, lg(n) in the worst. Wasn't sure it's worth clarifying that detail, but I certainly could if it helps.

File: internal/ethapi/api.go (hunk 5 above)
Comment (roberto-bayardo) [reply internal/ethapi/api.go]: yeah agree, I somehow missed the UsedGas value right there in the result :man_facepalming: 
fixed

File: internal/ethapi/api.go (hunk 2 above)
Comment (roberto-bayardo) [reply internal/ethapi/api.go:1230]: bit of both I guess: most txs don't need much higher gas limit than their gas used, and most txs don't require near the full block limit of gas. I guess this trick takes advantage of both facts. (FWIW EstimateGas doesn't seem to be called for simple 21k gas transfers by wallets like MetaMask so I excluded those from the analysis)

File: internal/ethapi/api.go (hunk 3 above)
Comment (roberto-bayardo) [reply internal/ethapi/api.go]: removed

File: internal/ethapi/api.go (hunk 1 above)
Comment (roberto-bayardo) [reply internal/ethapi/api.go]: done

**Comment by karalabe** [comment]
Whilst I understand that this code might make things a bit faster, but they also make it a lot more complicated. Do we have a specific use case where we need that extra drop of speed?

**Comment by karalabe** [comment]
Another interesting optimization we could consider is whether it's worth it to hone in on the *exact* gas usage needed, or if something "close enough" is enough.

I.e. By binary searching until `hi == lo + 1`, we're wasting a lot of executions to get the exact to-the-gas accuracy. Most wallets will just add some random number on top to account for network conditions. And I don't think it's relevant in the grand scheme of things when trying to fit txs into a block.

Lowering to accuracy from 1 gas to 1024 gas for example would save 10 iterations of the binary search.

**Comment by roberto-bayardo** [comment]
> Whilst I understand that this code might make things a bit faster, but they also make it a lot more complicated. Do we have a specific use case where we need that extra drop of speed?

My main goal was to improve the handling around handling tx revert, which contributed to a recent issue I encountered that ending up fixed [here](https://github.com/ethereum/go-ethereum/pull/27505.

Bulk of the change to this end is around moving the unconstrained execution up front instead leaving it to the end, which I felt makes the impl more understandable.

The additional complexity I did end up adding is ~4 LOC (L1228 initialization of lo, and the the midpoint adjustment in the main loop) -- seemed worth the savings given this is a very expensive call.  Figured RPC providers that make use of this would benefit.

**Comment by roberto-bayardo** [comment]
> Another interesting optimization we could consider is whether it's worth it to hone in on the _exact_ gas usage needed, or if something "close enough" is enough.
> 
> I.e. By binary searching until `hi == lo + 1`, we're wasting a lot of executions to get the exact to-the-gas accuracy. Most wallets will just add some random number on top to account for network conditions. And I don't think it's relevant in the grand scheme of things when trying to fit txs into a block.
> 
> Lowering to accuracy from 1 gas to 1024 gas for example would save 10 iterations of the binary search.

Agree that the "lowest possible" aspect of this function is probably unnecessary.

**Review by fjl — CHANGES_REQUESTED** [review]

**Comment by fjl** [comment]
It'd be nice to have the script you used to sample the transactions. That way we could also repeat your analysis.

File: internal/ethapi/api.go [hunk 6]
```diff
@@ -1086,34 +1086,12 @@ func DoCall(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash
...
-// revertError is an API error that encompasses an EVM revertal with JSON error
-// code and a binary data blob.
-type revertError struct {
-	error
-	reason string // revert reason hex encoded
-}
-
-// ErrorCode returns the JSON error code for a revertal.
-// See: https://github.com/ethereum/wiki/wiki/JSON-RPC-Error-Codes-Improvement-Proposal
-func (e *revertError) ErrorCode() int {
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply internal/ethapi/api.go:1110]: ahah, restored.

**Comment by roberto-bayardo** [comment]
> It'd be nice to have the script you used to sample the transactions. That way we could also repeat your analysis.

there is no standalone script unfortunately. the analysis was over logs from a hacked client which invoked both old & new implementations for each call and logged the # of iterations of both. I then added a hook in the txpool to call EstimateGas for each sampled transaction, and let it run over night before extracting the log data and dumping into a spreadsheet.

**Comment by antonydenyer** [comment]
There's a reasonably good test in besu that covers some funny edge cases with the 1/64th rule [for reference the results are here](https://github.com/hyperledger/besu/blob/main/acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/jsonrpc/EthEstimateGasAcceptanceTest.java#L54)

File: internal/ethapi/api.go [hunk 7]
```diff
@@ -1170,16 +1166,18 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
 	} else {
 		feeCap = common.Big0
 	}
+
+	state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	if state == nil || err != nil {
+		return 0, err
+	}
+	err = overrides.Apply(state)
+	if err != nil {
```
Comment (holiman) [inline internal/ethapi/api.go:1175]:
```suggestion
	if err := overrides.Apply(state); err != nil {
```
I nkow you just lifted it up, but still...

File: internal/ethapi/api.go (hunk 4 above)
Comment (holiman) [reply internal/ethapi/api.go:1229]: I think it's still basically a binary search, it divides up the range of possible outcomes every iteration, and discards one of the two subranges. The trick here is that the point where the range is divided is not on on the arithmetic middle. 

Some clarifying might be good, e.g. what you wrote below, something like 

> most txs don't need much higher gas limit than their gas used, and most txs don't require near the full block limit of gas, so the selection of where to bisect the range is skewed to favour the low side.

**Review by holiman — COMMENTED** [review]
LGTM, some minor nits, and I wonder if @fjl's comments were really addressed?

File: internal/ethapi/api.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply internal/ethapi/api.go:1110]: weird, let me try again.

[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply internal/ethapi/api.go]: I think I got it this time :/

File: internal/ethapi/api.go (hunk 7 above)
Comment (roberto-bayardo) [reply internal/ethapi/api.go]: done

**Comment by roberto-bayardo** [comment]
I have added the more detailed comment _"most txs don't need much higher gas limit than their gas used, and most txs don't require near the full block limit of gas, so the selection of where to bisect the range is skewed to favour the low side"_ where the midpoint is adjusted.

File: internal/ethapi/api.go (hunk 4 above)
Comment (roberto-bayardo) [reply internal/ethapi/api.go:1250]: added.

**Comment by fjl** [comment]
My concerns were adressed. I can't really comment on the binary search thing, seems alright though.

**Review by holiman — APPROVED** [review]
LGTM
